### ISSUE TITLE
maxima win64 --eval option

### ISSUE THREAD

**Issue by Mequin** [wxMaxima-developers/wxmaxima#1301]
Since several years, I have developed a schematic editor (for analog LTI circuits)
that sends to wxMaxima, by a simple push on an icon, a ".mac" file to evaluate through
its --eval option (and then remains open for further user interactive requests)

With the latest Maxima release 5.44.0, wxMaxima opens its window and hangs

Even calling wxMaxima from a dos shell, it hangs now when called with --eval file.mac

Regards,
   Jacques

**Comment by Mequin** [comment]
It seems that nobody is interested, or even concerned, by the fact that the  --eval  command line option of wxMaxima is broken since the latest Maxima release ( 5.44.0 ) 
This is unfortunate
This option is helpul to integrate wxMaxima in an overall automation flow

Regards,
   Jacques

PS:  for the time being I will stick with an older release, but I will resubmit an another ticket later on

**Comment by KubaO** [comment]
I’d suggest reopening the issue, then :) There certainly is concern, even if it’s not voiced. 

Closing and resubmitting tickets is counterproductive: there are multiple developers that get e-mailed each time an issue is closed or opened or otherwise changed. In other words: there is plenty of collective awareness, but a day still has only 24 hours.

I personally have several branches of work on wxMaxima I have to finish first, and this issue would be the next one on my plate after that if nobody else takes on it earlier. That’s about 1.5 months away at the moment, +/-2 weeks.

> It seems that nobody is interested, or even concerned, by the fact that the --eval command line option of wxMaxima is broken since the latest Maxima release ( 5.44.0 )

**Comment by gunterkoenigsmann** [comment]
I use tje --eval option daily. So we'll need to find out what is different between our uses.

**Comment by KubaO** [comment]
I also can’t reproduce it. **We need an example .mac file that illustrates the problem**.

**Comment by KubaO** [comment]
I also can’t reproduce it. **Mequin, we need the following:**

1. An example .mac file that illustrates the problem. You can redact it by replacing text with Lorem Ipsum.

2. The way you invoke Maxima - e.g. is it from a .bat file, some other script, and the exact arguments you pass to Maxima (I mean exact - you can redact local names but leave them in the same form, i.e. just replace the letters but leave other aspects like spaces or punctuation intact).

3. The contents of your environment. After opening the command prompt, type:

      set > env.txt

  Then edit the file to redact identifying information if needed, and attach it to this issue.

**Comment by Mequin** [comment]
I running the sbcl version on WINDOWS 10  64 bits 

The "hello.mac" file only contains a single line :     print( "Hello World" ) $

To make the example very simple, I am launching wxMaxima directly from DOS 

With Maxima release 5.43.2,  wxMaxima opens and diplay "Hello World"

       C:\maxima-5.43.2-win64\wxmaxima\wxMaxima.exe --eval D:\jm\home\lab\hello.mac

But with Maxima release 5.44.0  wxMaxim opens and hangs without even showing a prompt

       C:\maxima-5.44.0-win64\bin\wxMaxima.exe      --eval D:\jm\home\lab\hello.mac

build_info("5.44.0","2020-06-08 22:46:53","x86_64-w64-mingw32","SBCL","2.0.0",
"D:/JM/HOME/.maxima","D:/Utilisateurs/Browny/AppData/Local/Temp",
"D:/JM/HOME/.maxima/binary/5_44_0/sbcl/2_0_0",
"wxMaxima","20.06.6_MSW_OpenMP201511+Locks")"

Regards,
  Jacques

PS: Just in case, I have disabled my antivirus

**Comment by gunterkoenigsmann** [comment]
Thanks a lot! It's midnight in germany. But today I will try to reproduce your setup on my linux computer. My hope is that we have accidentally resolved the error since the last wxMaxima release. In that case uninstalling wxMaxima and installing the one from https://wolfgang.dautermann.at/maxima/nightlybuild/ instead should resolve the problem.

**Comment by gunterkoenigsmann** [comment]
Antivirusses are a difficult theme on MS Windows: For some reason on windows moving, renaming or deleting files is impossible while they are still scanned by a virus scanner - even if the whole framework for allowing to do so seems to be in place: For example all that has to be done if a file is to be renamed is to change the directory entry for the file. The file itself stays where it was during that operation.

wxMaxima tries to ship around this kind of problems in two ways:

- It tries to use as low a number of external files as possible, and
- if a rename, delete or move operation fails it retries one, two and three seconds later hoping that by then the virus scanner will have finished. If that still isn't the case then wxMaxima will inform the user or decide that deleting temp files hasn't been a high-prio job, anyway.

If you find something a virus scanner is able to break: I cannot guarantee to manage to address this kind of issue fast. Things I cannot reproduce on my computer are always a bit of guesswork. But I can at least state that I try to fix this kind of issue.

**Comment by Mequin** [comment]
Just to be clear, I have uninstalled all previous Maxima release
I have deleted the wxMaxima entry in my Windows 10 registry

              >    C:\maxima-5.44.0-win64\bin\wxmaxima.exe     WORKS FINE
                    desktop windows is open and prompt waiting for user interraction
but 
              >   C:\maxima-5.44.0-win64\bin\wxmaxima.exe  --eval D:\JM\HOME\LAB\hello.mac     FAIILS TO WORK
                    desktop windows is open 
                    but "Hello World" is not displayed and there is no prompt waiting for further user interraction
  
So,  I do not think this problem related with my antivirus nor ".bat' file since things are fine without  the   --eval option

As always, thanks for your help

Regards,
    Jacques

**Comment by Mequin** [comment]
[sarc.pdf](https://github.com/wxMaxima-developers/wxmaxima/files/4868241/sarc.pdf)

This PDF file is just to show you how critical the --eval option is for my application

Best regards,
    Jacques

**Comment by gunterkoenigsmann** [comment]
Your application looks great!

...and a tester who has the ability to trigger a bug is a valuable resource: Without being able to trigger a bug there is nearly no way to find the code line that triggers it.

But the PDF file you have sent me might be a valuable hint: Might it be that the problem will be triggered even if you omit the `--eval` from the command line? The background to that question is: The wxMaxima version that was shipped with maxima 5.44.0 crashes on opening certain .mac files. If that is the problem using the wxMaxima version that can be found at https://wolfgang.dautermann.at/maxima/nightlybuild/ instead resolves it.

**Comment by Mequin** [comment]
I am able to execute the "mac" file generated by my application directly in wxMaxima ( without passing by --eval )

Moreover, the problem also occurs with a trivial mac file containing only    print( "Hello World" ) $

I will look at your web link during the weekend

Regards

**Comment by KubaO** [comment]
Could you please attach the output of the SYSTEMINFO command to this report?

In command line, run

SYSTEMINFO > info.csv

Then open it in your favorite editor, redact for privacy, and attach.

Also, have you tried with the nightly build? I can open all mac files included as tests with Maxima using my developer build. The earliest I can properly look at it is Monday evening.

I also have to run tests on our Windows CI, so if there would be such a problem, it would be caught. But given that the developer builds show no problems, I’m not holding my horses just yet.

**Comment by Mequin** [comment]
I have downloaded the latest nightly build   ( maxima-clisp-sbcl-abcl-current-win64.exe  2020-07-04 00:51  153M   ) 

IT WORKS   !!!!

Thanks for your time, I guess that now the ticket can be really closed

I have joined a PDF to explain what I am doing with Maxima/wxMaxima

Best Regards,
    Jacques Mequin

![screenshot](https://user-images.githubusercontent.com/14158309/86516839-31185600-be24-11ea-9c17-6315f2c188ae.png)
[SARC.pdf](https://github.com/wxMaxima-developers/wxmaxima/files/4873192/SARC.pdf)

**Comment by Mequin** [comment]
So, the option --eval  is back on business

However, executing my code,  with or without this option,  is much less fluid than with older version
Sometimes, the screen seems to wait forever before displaying something

But this is probably unrelated problem

Regards,
     Jacques

I will stick with 5.43.2 for the time being

**Comment by gunterkoenigsmann** [comment]
If you can provide us with a file that has grown slower we'll see what we can do: Most of the work of the last few months was speeding up the display code. Not that one such attempt backfired for your use-case...

**Comment by Mequin** [comment]
This run is made without the --eval option by pasting a (small ) code directly on the screen 
In fact the wxMaxima windows spent most of the displaying "analyse du resulat" in its lower right corner

A job taking less than a minute with 5.43.2 takes more than 6 minutes with the nightly build
By the way, adding "the time spent in Maxima"  I get only around 28 secondes the reste is display
So 5.43.2 is not doing too bad

I will see what I can do to send you something but it is not going to be simple to package someting
[run_wxMaxima.zip](https://github.com/wxMaxima-developers/wxmaxima/files/4873587/run_wxMaxima.zip)

Regards,
   Jacques

**Comment by KubaO** [comment]
That PDF you show is a little gem. Beautiful.

**Comment by KubaO** [comment]
There are performance regressions against old versions, and they’ll take a while to eradicate, but there are lots of very low hanging fruit that I know how to take care of. They stand out in profiles like sore thumbs.

**Comment by Mequin** [comment]
In fact the previous kind of distibution was more convenient
I remember that is was simple to install the new release of Maxima and just rename its "wxMaxima folder" to put an 
older one instead
I have used that trick several times in the past
Now that wxMaxima is merge with the distribution it is not as simple
Regards

**Comment by gunterkoenigsmann** [comment]
In theory the MinGW flavour of the appveyor build should be a standalone wxMaxima application...

**Comment by KubaO** [comment]
Indeed. I just didn’t check of they may be debug builds, and they should also be statically linked to the C runtime. Then wxMaxima won’t require installation - just unzip the binaries. But the installer of course is generated anyway.

**Comment by gunterkoenigsmann** [comment]
In theory we could tell AppVeyor to offer to download the .exe file as Artifact.

**Comment by Mequin** [comment]
Again, thanks for all your comments

Since the  --eval  option is back on business you can probably close this ticket

I will propably open a new ticket about the "speed reduction" after packaging for you
some demonstrative data 
Clearly something goes wrong when a job take 48s on 5.43.2 and 6m18s on 5.44.0
and that prints about the "runtime time spent in Maxima is only 28s
Moreover, in that job, lines and matrices were far from very large

I am only doing "print" and "disp" and no fancy feature ( image, plot, or font change)
In fact I am mostly using "Courrier New" that is better readable 

As a 25 years Maxima/wxMaxima user I know that they can have bugs at different time

Last years I did have a "fasl" compilation problem with sbcl Maxima
Being able to plug, in the previous version of Maxima, the working wxMaxima folder
was an easy workaround until the bug fix  ( and the other way around )

Regards,
   Jacques

**Comment by Mequin** [comment]
Just to verify, I have re-executed the experiment with a chronometer

with 5.43.2 

     total execution time  41 s

     Maxima runtime   8s69 + 13s75 + 14s75 = 37s19

     => display time around 4 secondes

with nightly build ( July 4 )

      total execution time 6 minutes 25s

      Maxima runtime   8s69 + 13s6 + 14s62 = 36s37 (almost the same than above)

      ==> display time  5 minutes 49 secondes 

I am joining the ".wxmx"

Regards
   Jacques

[untitled.wxmx.zip](https://github.com/wxMaxima-developers/wxmaxima/files/4875060/untitled.wxmx.zip)

**Comment by gunterkoenigsmann** [comment]
Thanks a lot!

If we draw part of the cells massively more than once for every update of the worksheet?

**Comment by KubaO** [comment]
One thing that takes horrible amount of time is recalculation on every groupcell change (adding a cell). This affects opening/loading the file.

But with a profiler it will be obvious what’s wrong and may even be a simple patch to get rid of the quadratic behavior.

**Comment by Mequin** [comment]
Let me explain the ".wxmx" file 

UP TO the line  /* EOF */   this corresponds to the data sent by my editor with the  --eval  option

NETLIST(  )  OUTPUTS(  )  GENERATE_MIMO(  )  GENERATE_Z(  )  are autoload Maxima definition

Thus, the associated code is executed

As a consequence, since it is a fully  automated flow, all the display operations ( print , disp ) are defined
within the various "load" files 

It should be noted that I am sometimes using "disp" to force the update of the screen in order
to monitor the progression of the flow  ( somewhat a "flush"  forcing pending "print"  )

I hope this help

Regards,
     Jacques

**Comment by gunterkoenigsmann** [comment]
Filed issue #1326 so we have a separate issue we can work on for the performance issue. @Mequin : If you want to be informed about that issue, too, I believe you have to star it or similar.

**Comment by KubaO** [comment]
I'll close this for now. Speed issues would be best reported under #1326.
